draft-bingler-hsts-tracking-prevention.md

```diff
@@ -0,0 +1,245 @@
+
+HTTP Strict Transport Security (HSTS), {RFC6797}, improves security for users
+by allowing a host to tell the user agent that it should only ever be accessed
+over a secure connection.
+
+However, the design of HSTS allows for servers to abuse its state caching in
+order to set and retrieve arbitrary data, potentially for the purposes of
+tracking a user around the web. {Section 14.9 of RFC6797} warns of this
+possibility but does not discuss prevention.
+
+This document offers some minor modifications to {Section 8 of RFC6797}
+that allows for user agents to implement a way to prevent tracking of the user.
+
+It also touches on potential implementations and their downsides.
+
+# Conventions and Definitions
+
+{::boilerplate bcp14-tagged}
+
+# HSTS Tracking Prevention
+
+Tracking prevention can take a couple of forms:
+
+* Prevent servers from setting tracking data.
+
+* Prevent servers from retrieving tracking data.
+
+* Make correlating data between sites infeasible.
+
+The first could be implemented by ignoring Strict-Transport-Security response
+headers from tracking hosts. This requires user agents to ignore otherwise
+valid Strict-Transport-Security response headers if the user agent has reason
+to believe that header would be used for tracking purposes.
+
+The second could be implemented by not upgrading URI for tracking hosts.
+This requires user agents to not upgrade otherwise eligible URIs to
+"https" if the user agent has reason to believe the upgraded URI would be used
+for tracking purposes.
+
+The third could be implemented by partitioning the HSTS policy store by the
+current site the user is on. This would force any server interested in tracking
+a user to set new data on every site, making it infeasible to correlate the
+user’s visits between sites via HSTS upgrades.
+
+## Monkey-Patches against RFC6797
+
+### Prevent Servers From Saving Tracking Data
+
+To allow user agents the option of preventing hosts they believe are tracking
+the user from setting HSTS data {Section 8.1.1 of RFC6797} should be changed
+as follows:
+
+~~~
+   If the substring matching the host production from the Request-URI
+   (of the message to which the host responded) syntactically matches
+   the IP-literal or IPv4address productions from Section 3.2.2 of
+   [RFC3986], then the UA MUST NOT note this host as a Known HSTS Host.
+
+   Otherwise, if the substring...
+~~~
+
+should be modified to include
+
+~~~
+   If the substring matching the host production from the Request-URI
+   (of the message to which the host responded) syntactically matches
+   the IP-literal or IPv4address productions from Section 3.2.2 of
+   [RFC3986], then the UA MUST NOT note this host as a Known HSTS Host.
+
+   If the UA has reason to believe this host intends to set tracking
+   data then the UA MAY not note this host as a Known HSTS Host.
+
+   Otherwise, if the substring...
+~~~
+
+{Section 2.4.1.1 of RFC6797} step 4 may also need to be rewritten in a similar
```

Comment:
Not having read it, why is this only a may? Can we just add the monkey patch here to be explicit?

Reply:
I think I'm just going to remove that sentence. It's really just me musing outloud rather than anything instructive.